File: engine/belgique.py
"""Calculs fiscaux — Belgique (revenus 2025 / EI 2026).

Couvre : IPP (barème + quotité exemptée + additionnels communaux),
cotisations sociales indépendant, ISOC.
"""

from __future__ import annotations

from .core import round2, round4, tax_from_brackets
from .loader import load_country

_D = lambda: load_country("BE")  # noqa: E731


def ipp(revenu_imposable: float, taux_communal: float = 0.0) -> dict:
    """Impôt des personnes physiques.

    Args:
        revenu_imposable: revenu imposable (€).
        taux_communal: taux des additionnels communaux (ex. 0.08 pour 8 %).
    """
    p = _D()["personnes_physiques"]["impot_revenu_ipp"]
    impot_bareme = tax_from_brackets(revenu_imposable, p["tranches"])

    # Quotité exemptée : exonérée au taux de base (1re tranche)
    taux_base = p["tranches"][0]["taux"]
    reduction_quotite = p["quotite_exemptee_base"] * taux_base
    impot_apres_quotite = max(0.0, impot_bareme - reduction_quotite)

    additionnels = impot_apres_quotite * taux_communal
    total = impot_apres_quotite + additionnels
    return {
        "impot_bareme": round2(impot_bareme),
        "reduction_quotite": round2(reduction_quotite),
        "impot_apres_quotite": round2(impot_apres_quotite),
        "additionnels_communaux": round2(additionnels),
        "impot_total": round2(total),
    }


def cotisations_independant(revenu_net: float) -> dict:
    """Cotisations sociales INASTI (titre principal), par paliers."""
    paliers = _D()["independants"]["cotisations_inasti"]["paliers"]
    cot = tax_from_brackets(revenu_net, paliers)
    return {
        "cotisations_annuelles": round2(cot),
        "cotisations_trimestrielles": round2(cot / 4),
    }


def impot_societes(benefice: float, pme_taux_reduit: bool = True) -> dict:
    """Impôt des sociétés (ISOC)."""
    s = _D()["societes"]["isoc"]
    plafond = s["plafond_taux_reduit"]
    if pme_taux_reduit and benefice > 0:
        part_reduite = min(benefice, plafond)
        part_normale = max(0.0, benefice - plafond)
        impot = part_reduite * s["taux_reduit"] + part_normale * s["taux_normal"]
    else:
        impot = benefice * s["taux_normal"]
    return {"impot": round2(impot), "taux_effectif": round4(impot / benefice) if benefice else 0.0}

File: src/belgique.ts
/**
 * Calculs fiscaux — Belgique (revenus 2025 / EI 2026).
 * IPP (barème + quotité + additionnels communaux), cotisations indépendant, ISOC.
 */

import { round2, round4, taxFromBrackets, Bracket } from "./core.js";
import { loadCountry } from "./loader.js";

const D = () => loadCountry("BE");

/** Impôt des personnes physiques. */
export function ipp(revenuImposable: number, tauxCommunal = 0) {
  const p = D().personnes_physiques.impot_revenu_ipp;
  const tranches = p.tranches as Bracket[];
  const impotBareme = taxFromBrackets(revenuImposable, tranches);

  const tauxBase = tranches[0].taux;
  const reductionQuotite = p.quotite_exemptee_base * tauxBase;
  const impotApresQuotite = Math.max(0, impotBareme - reductionQuotite);

  const additionnels = impotApresQuotite * tauxCommunal;
  return {
    impot_bareme: round2(impotBareme),
    reduction_quotite: round2(reductionQuotite),
    impot_apres_quotite: round2(impotApresQuotite),
    additionnels_communaux: round2(additionnels),
    impot_total: round2(impotApresQuotite + additionnels),
  };
}

/** Cotisations sociales INASTI (titre principal), par paliers. */
export function cotisationsIndependant(revenuNet: number) {
  const paliers = D().independants.cotisations_inasti.paliers as Bracket[];
  const cot = taxFromBrackets(revenuNet, paliers);
  return {
    cotisations_annuelles: round2(cot),
    cotisations_trimestrielles: round2(cot / 4),
  };
}

/** Précompte mobilier (dividendes/intérêts). Taux par défaut = régime général (30 %). */
export function precompteMobilier(montant: number, taux?: number) {
  const p = D().personnes_physiques.precompte_mobilier;
  const t = taux ?? p.dividendes_general;
  const impot = montant * t;
  return { taux: t, impot: round2(impot), net: round2(montant - impot) };
}

/** Impôt des sociétés (ISOC). */
export function impotSocietes(benefice: number, pmeTauxReduit = true) {
  const s = D().societes.isoc;
  const plafond = s.plafond_taux_reduit;
  let impot: number;
  if (pmeTauxReduit && benefice > 0) {
    const partReduite = Math.min(benefice, plafond);
    const partNormale = Math.max(0, benefice - plafond);
    impot = partReduite * s.taux_reduit + partNormale * s.taux_normal;
  } else {
    impot = benefice * s.taux_normal;
  }
  return { impot: round2(impot), taux_effectif: benefice ? round4(impot / benefice) : 0 };
}
